### PR TITLE
go.mod: change imports to github.com/distribution/distribution/v3

### PR DESCRIPTION
**Pull request by thaJeztah**
Go 1.13 and up enforce import paths to be versioned if a project
contains a go.mod.

The current v2.x branches (and releases) do not yet have a go.mod,
and therefore are still allowed to be imported with a non-versioned
import path (go modules add a `+incompatible` annotation in that case).

However, now that this project has a `go.mod` file, incompatible
import paths will not be accepted by go modules, and attempting
to use code from this repository will fail.

This patch uses `v3` for the import-paths (not `v2`), because changing
import paths itself is a breaking change, which means that  the
next release should increment the "major" version to comply with
SemVer (as go modules dictate).

### CODE REVIEW

**Comment by thaJeztah** [comment]
@manishtomar @olegburov @dmcgowan @stevvooe PTAL: adding go modules in this repository forces us to update the import paths (because this repository already has `v2.xx` tags); because of that, the next release must be a v3.xx release (unfortunately, go modules doesn't allow separating binary from golang releases for the whole repo)

**Comment by thaJeztah** [comment]
Note that the recommendation is to move all code to a `vXX` subdirectory to support vendoring tools that do not understand versioned import-paths; I decided not to make that change (because this project uses release branches)

File: version/version.go
Comment (thaJeztah) [inline version/version.go:11]: Just noticed this version string; updated to `v3.0.0` to match the change

File: Makefile
Comment (thaJeztah) [inline Makefile:9]: Missed this one, which caused some paths not being excluded, what caused CI to fail.

all green now

**Comment by olegburov** [comment]
> This patch uses v3 for the import-paths (not v2), because changing import paths itself is a breaking change, which means that the next release should increment the "major" version to comply with

If there's still a plan for [2.x release(s)](https://github.com/docker/distribution/milestones), maybe it make sense to merge this after and start `3.x` releases.

**Comment by thaJeztah** [comment]
> If there's still a plan for 2.x release(s), maybe it make sense to merge this after and start 3.x releases.

There's release/2.x branches, so for changes to go into a 2.x release, I guess those will have to be cherry picked into those branches

**Comment by thaJeztah** [comment]
Rebased this one

**Comment by thaJeztah** [comment]
With this repository being moved to https://github.com/distribution/distribution, we should probably rename the module (and imports) accordingly, further qualifying next release to be v3.0.0

**Comment by milosgajdos** [comment]
Whilst I'm waiting for the GH org invite so I can start GH review, @thaJeztah could you please change the `import` paths to the new GH org URL?

**Comment by thaJeztah** [comment]
Yup; will update the PR to use the new home 👍

**Comment by thaJeztah** [comment]
Updated 👍

File: go.mod
Comment (milosgajdos) [inline go.mod:3]: Should we bump this to `1.13` at least? Or ideally, match it with what's in `travis.yaml` i.e. `1.14`?

Comment (thaJeztah) [reply go.mod:3]: I was considering changing it, but it's not directly related to this change. Generally I try to be a conservative to allow projects to build with other Golang versions (although Golang does not (yet) "enforce" the go version, [it's possible they start doing so](https://github.com/golang/go/issues/30446#issuecomment-468044008)).

Comment (joaodrp) [reply go.mod:3]: This should probably reflect the minimum required version to build the project. I don't think we rely in any 1.13+ feature, so seems OK to keep it _as-is_.

Comment (thaJeztah) [reply go.mod:3]: Yes; and even if building the "whole" project requires Go version X, it's possible there are consumers that only use a single package from this repo, which may in itself not require that version; we could _document_ that the project was "tested on" / "requires" / "recommends" a specific Go version to build (e.g., in "BUILDING.md").

Comment (milosgajdos) [reply go.mod:3]: 👍 Sounds good. I was thinking more along the lines of "dropping the support for older versions of Go", but this is ok with me. Adding `BUILD.md` file with notes and instructions sounds good.

**Review by milosgajdos — APPROVED** [review]

**Review by joaodrp — APPROVED** [review]

**Comment by wy65701436** [comment]
Can we merge it? Then the dependencies can update their source.

**Comment by joaodrp** [comment]
> Can we merge it? Then the dependencies can update their source.

I think so, it just needs a rebase.

**Comment by thaJeztah** [comment]
Rebased, and did a quick find & replace to check if I missed renames from docker/distribution to distribution/distribution

**Comment by milosgajdos** [comment]
We are in a process of cutting a patch release for `2.7.x`, but we are still sorting out some details.
There is no immediate plan for `3.x` release I'm afraid. Maintainers need to agree on the roadmap, but first we need to cut the `2.7.x` release. Stay tuned @RainbowMango

**Comment by RainbowMango** [comment]
Thanks @milosgajdos. Good to know.

**Pull request merged**
